client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,366 @@
+	accum.startTime, accum.endTime = time.Time{}, time.Time{}
+	accum.topoMap = nil
+}
+
+func (accum *fetchTaggedResultAccumulator) Reset(
+	startTime time.Time,
+	endTime time.Time,
+	topoMap topology.Map,
+	majority int,
+	consistencyLevel topology.ReadConsistencyLevel,
+) {
+	accum.exhaustive = true
+	accum.startTime = startTime
+	accum.endTime = endTime
+	accum.topoMap = topoMap
+	accum.majority = majority
+	accum.consistencyLevel = consistencyLevel
+	accum.numHostsPending = int32(topoMap.HostsLen())
+	accum.numShardsPending = int32(len(topoMap.ShardSet().All()))
+
+	// expand shardResults as much as necessary
+	targetLen := 1 + int(topoMap.ShardSet().Max())
+	accum.shardResults = fetchTaggedShardConsistencyResults(accum.shardResults).initialize(targetLen)
+	// initialize shardResults based on current topology
+	for _, hss := range topoMap.HostShardSets() {
+		for _, hShard := range hss.ShardSet().All() {
+			id := int(hShard.ID())
+			accum.shardResults[id].enqueued++
+		}
+	}
+}
+
+func (accum *fetchTaggedResultAccumulator) sliceResponsesAsSeriesIter(
+	pools fetchTaggedPools,
+	elems fetchTaggedIDResults,
+) encoding.SeriesIterator {
+	numElems := len(elems)
+	iters := pools.IteratorArray().Get(numElems)[:numElems]
+	for idx, elem := range elems {
+		slicesIter := pools.ReaderSliceOfSlicesIterator().Get()
+		slicesIter.Reset(elem.Segments)
+		multiIter := pools.MultiReaderIterator().Get()
+		multiIter.ResetSliceOfSlices(slicesIter)
+		iters[idx] = multiIter
+	}
+
+	// pick the first element as they all have identical ids/tags
+	// NB: safe to assume this element exists as it's only called within
+	// a forEachID lambda, which provides the guarantee that len(elems) != 0
+	elem := elems[0]
+
+	encodedTags := pools.CheckedBytesWrapper().Get(elem.EncodedTags)
+	decoder := pools.TagDecoder().Get()
+	decoder.Reset(encodedTags)
+
+	tsID := pools.CheckedBytesWrapper().Get(elem.ID)
+	nsID := pools.CheckedBytesWrapper().Get(elem.NameSpace)
+	seriesIter := pools.SeriesIterator().Get()
+	seriesIter.Reset(pools.ID().BinaryID(tsID), pools.ID().BinaryID(nsID),
+		decoder, accum.startTime, accum.endTime, iters)
+
+	return seriesIter
+}
+
+func (accum *fetchTaggedResultAccumulator) AsEncodingSeriesIterators(
+	limit int, pools fetchTaggedPools,
+) (encoding.SeriesIterators, bool, error) {
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+
+	numElements := 0
+	accum.responses.forEachID(func(_ fetchTaggedIDResults) bool {
+		numElements++
+		return numElements < limit
+	})
+
+	result := pools.MutableSeriesIterators().Get(numElements)
+	result.Reset(numElements)
+	count := 0
+	accum.responses.forEachID(func(elems fetchTaggedIDResults) bool {
+		seriesIter := accum.sliceResponsesAsSeriesIter(pools, elems)
+		result.SetAt(count, seriesIter)
+		count++
+		return count < limit
+	})
+
+	exhaustive := accum.exhaustive && count < limit
```

Comment:
Should this be `count <= limit`?  If so, maybe we should add a "test off by one" test case for this? (I know its such a minor issue.. shrug)

Reply:
fair, will do.

Reply:
done